parser/src/infix.rs

```diff
@@ -0,0 +1,413 @@
+
+use base::ast::{Expr, IdentEnv, SpannedExpr, TypedIdent};
+use base::fnv::FnvMap;
+use std::cmp::Ordering;
+use std::ops::Index;
+
+#[derive(Copy, Clone, Debug)]
+pub enum Fixity {
+    /// Left operator associativity.
+    ///
+    /// For example, when the `(~)` operator is left-associative:
+    ///
+    /// ```text
+    /// x ~ y ~ z ≡ (x ~ y) ~ z
+    /// ```
+    Left,
+    /// Right operator associativity.
+    ///
+    /// For example, when the `(~)` operator is right-associative:
+    ///
+    /// ```text
+    /// x ~ y ~ z ≡ x ~ (y ~ z)
+    /// ```
+    Right,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct Operator {
+    pub precedence: i32,
+    pub fixity: Fixity,
+}
+
+impl Operator {
+    fn new(precedence: i32, fixity: Fixity) -> Operator {
+        Operator {
+            precedence: precedence,
+            fixity: fixity,
+        }
+    }
+}
+
+pub struct OperatorTable {
+    pub operators: FnvMap<String, Operator>,
+    pub default: Operator,
+}
+
+impl OperatorTable {
+    fn new<I>(ops: I) -> OperatorTable
+        where I: IntoIterator<Item = (&'static str, Operator)>,
+    {
+        OperatorTable {
+            operators: ops.into_iter().map(|(name, op)| (name.to_string(), op)).collect(),
+            default: Operator::new(9, Fixity::Left),
+        }
+    }
+}
+
+impl Default for OperatorTable {
+    fn default() -> OperatorTable {
+        OperatorTable::new(vec![
+            ("*", Operator::new(7, Fixity::Left)),
+            ("/", Operator::new(7, Fixity::Left)),
+            ("%", Operator::new(7, Fixity::Left)),
+            ("+", Operator::new(6, Fixity::Left)),
+            ("-", Operator::new(6, Fixity::Left)),
+            (":", Operator::new(5, Fixity::Right)),
+            ("++", Operator::new(5, Fixity::Right)),
+            ("&&", Operator::new(3, Fixity::Right)),
+            ("||", Operator::new(2, Fixity::Right)),
+            ("$", Operator::new(0, Fixity::Right)),
+            ("==", Operator::new(4, Fixity::Left)),
+            ("/=", Operator::new(4, Fixity::Left)),
+            ("<", Operator::new(4, Fixity::Left)),
+            (">", Operator::new(4, Fixity::Left)),
+            ("<=", Operator::new(4, Fixity::Left)),
+            (">=", Operator::new(4, Fixity::Left)),
+            // Hack for some library operators
+            ("<<", Operator::new(9, Fixity::Right)),
+            (">>", Operator::new(9, Fixity::Left)),
+            ("<|", Operator::new(0, Fixity::Right)),
+            ("|>", Operator::new(0, Fixity::Left)),
+        ])
+    }
+}
+
+impl<'a> Index<&'a str> for OperatorTable {
+    type Output = Operator;
+
+    fn index(&self, name: &'a str) -> &Operator {
+        self.operators.get(name).unwrap_or_else(|| {
+            if name.starts_with("#") {
+                &self[name[1..].trim_left_matches(char::is_alphanumeric)]
+            } else {
+                &self.default
+            }
+        })
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum Error {
```

Comment:
:+1: Needs a `std::fmt::Display` and maybe `std::error::Error` impl as well. Adding the operators involved in the conflict to the error would be nice as well.